### ISSUE TITLE
Failure to build on OS X 10.11 with Xcode 8

### ISSUE THREAD

**Issue by ararslan** [cisco/ChezScheme#146]
Chez Scheme version 9.4 fails to build on OS X 10.11 (El Capitan) with Xcode 8, since `_clock_gettime` isn't found. Xcode 8 defines certain symbols to be able to cross-compile for Sierra, but doesn't actually make them available on the local system. Adding `LDFLAGS=-Wl,-no_weak_imports` to `configure` allows the build to succeed; perhaps that could be made the default in this case?

**Comment by akeep** [comment]
I've been trying to reproduce this problem, but I've had some trouble getting it to fail the way you are seeing it fail.  I've upgraded Xcode on my El Capitan install, however, I'm still able to build and test the scheme compiler without any extra flags.  I've had this problem in the past with upgrades to macOS and Xcode seemingly leaving some previous settings around, so I'll try to see if I can resolve this.

I'm a bit concerned that if we use linker flags to solve this problem it may lead to run-time problems, if we are ending up with the ``_clock_gettime`` symbol left unbound.

I took a look through the source code, and it looks like we decide to use the ``clock_gettime`` function in our statistics package only when certain C macros are defined, and fall back to something else in other cases.  I think we might need to augment this with some special checking for the version of macOS we are running on to use the fall back on El Capitan.

The macros we are checking for are: ``CLOCK_THREAD_CPUTIME_ID``, ``CLOCK_PROCESS_CPUTIME_ID``, ``CLOCK_MONOTONIC_HR``, ``CLOCK_MONOTONIC``, ``CLOCK_HIGHRES``, ``CLOCK_REALTIME_HR``, and ``CLOCK_REALTIME``.  When these macros are not defined in an included header file, it falls through to use either ``getrusage`` or ``gettimeofday``, depending on the type of time that is being requested.  See the ``s_gettime`` function in ``stats.c`` for the details.

Anyway, I wanted to let you know that I'm looking into it and try to see if I can get something I can test worked out.

**Comment by ararslan** [comment]
Hey @akeep, thanks so much for looking into this! Perhaps a possible solution would be to ignore `clock_gettime` entirely on versions of macOS prior to Sierra (10.12), since that's when it was actually included with the system. How about something like:
```c
#ifdef __APPLE__
#include <AvailabilityMacros.h>

#if MAC_OS_X_VERSION_MIN_REQUIRED >= MAC_OS_X_VERSION_10_12
    #define HAS_CLOCK_GETTIME 1
#else
    #define HAS_CLOCK_GETTIME 0
#endif
#endif
```
Then anything here that relies on `clock_gettime` being available can be wrapped in `#if HAS_CLOCK_GETTIME`. Does something like that seem reasonable?

**Comment by akeep** [comment]
Yes, this is what I had in mind.  We'd probably want to bake this ``HAS_CLOCK_GETTIME`` into the ``system.h`` so that it is consistent across supported operating systems, and put the special handling for the macOS case into the ``*osx`` machine type sections of ``system.h``.

**Comment by okuoku** [comment]
Hey @ararslan , have you installed "Command Line Tools (mac OS 10.11) for Xcode 8.2"? (it's available on https://developer.apple.com/download/more/ )

(Surprisingly,) Xcode8 does not come with SDK for 10.11 even if it was installed on 10.11 system. So you will need this separate package. Note that Xcode 8.2 is the last version that supports macOS 10.11.

**Comment by ilovezfs** [comment]
>(Surprisingly,) Xcode8 does not come with SDK for 10.11 even if it was installed on 10.11 system. So you will need this separate package.

This is not correct. Xcode 8 is sufficient without the CLT. The "10.12" is just a symlink to the unitary SDK, which is intended for use on 10.11 and 10.12. You set the MACOSX_DEPLOYMENT_TARGET accordingly.
